File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/cf/MsiFile.kt
package com.jetbrains.signatureverifier.cf

import com.jetbrains.signatureverifier.SignatureData
import com.jetbrains.signatureverifier.serialization.toHexString
import org.jetbrains.annotations.NotNull
import java.nio.channels.SeekableByteChannel
import java.security.MessageDigest

/**
 * MS Windows Installer compound file
 */
open class MsiFile {
  private val _cf: CompoundFile
  val fileSize: Long

  companion object {
    //\u0005DigitalSignature
    val digitalSignatureEntryName = arrayOf<Byte>(
      0x5, 0x0, 0x44, 0x00, 0x69, 0x00, 0x67, 0x00, 0x69, 0x00, 0x74, 0x00, 0x61, 0x00, 0x6C, 0x00, 0x53,
      0x00, 0x69, 0x00, 0x67, 0x00, 0x6E, 0x00, 0x61, 0x00, 0x74, 0x00, 0x75, 0x00, 0x72, 0x00, 0x65, 0x00
    ).toByteArray()

    //\u0005MsiDigitalSignatureEx
    val msiDigitalSignatureExEntryName = arrayOf<Byte>(
      0x5, 0x0, 0x4D, 0x00, 0x73, 0x00, 0x69, 0x00, 0x44, 0x00, 0x69, 0x00, 0x67, 0x00, 0x69, 0x00, 0x74, 0x00, 0x61,
      0x00, 0x6C, 0x00, 0x53, 0x00, 0x69, 0x00, 0x67, 0x00, 0x6E, 0x00, 0x61, 0x00, 0x74, 0x00, 0x75, 0x00, 0x72, 0x00,
      0x65, 0x00, 0x45, 0x00, 0x78, 0x00
    ).toByteArray()

    val rootEntryName = arrayOf<Byte>(
      0x52, 0x00, 0x6F, 0x00, 0x6F, 0x00, 0x74, 0x00, 0x20, 0x00, 0x45, 0x00, 0x6E, 0x00, 0x74, 0x00, 0x72, 0x00,
      0x79, 0x00
    ).toByteArray()

    val hexNamesSet = setOf(
      digitalSignatureEntryName.toHexString(),
      msiDigitalSignatureExEntryName.toHexString(),
      rootEntryName.toHexString()
    )
  }

  /**
   * Initializes a new instance of the MsiFile
   *
   * @param stream  An input stream
   * @exception PlatformNotSupportedException  Indicates the byte order ("endianness")
   *      in which data is stored in this computer architecture is not Little Endian.
   * @exception InvalidDataException  If the input stream contains a compound file with wrong structure
   */
  constructor(@NotNull stream: SeekableByteChannel) {
    _cf = CompoundFile(stream)
    fileSize = stream.size()
  }

  /***
   * Initializes a new instance of the MsiFile from compound file json dump
   */
  constructor(compoundFileMetaInfo: CompoundFile.Companion.CompoundFileMetaInfo, stream: SeekableByteChannel) {
    _cf = CompoundFile(compoundFileMetaInfo, stream)
    fileSize = stream.size()
  }

  fun getCFMetaInfo() = _cf.getMetaInfo()

  fun getEntries(visitedSectors: MutableList<Pair<Int, Int>>? = null) = _cf.getEntries(visitedSectors)
  fun getRootEntry(visitedSectors: MutableList<Pair<Int, Int>>? = null) = _cf.getRootEntry(visitedSectors)

  fun putEntries(
    data: List<Pair<DirectoryEntry, ByteArray>>,
    miniStreamStartSector: Int,
    wipe: Boolean = false
  ) =
    _cf.putEntries(data, miniStreamStartSector, wipe)

  /**
   * Retrieve the signature data from MSI
   */
  fun GetSignatureData(): SignatureData {
    val data = _cf.GetStreamData(digitalSignatureEntryName)

    if (data == null)
      return SignatureData.Empty

    return SignatureData(null, data)
  }

  /**
   * Compute hash of MSI structure
   *
   * @param algName  Name of the hashing algorithm
   * @param skipMsiDigitalSignatureExEntry  Skip \u0005MsiDigitalSignatureEx entry data when hashing
   */
  fun ComputeHash(@NotNull algName: String, skipMsiDigitalSignatureExEntry: Boolean): ByteArray {
    val entries = _cf.GetStreamDirectoryEntries()
      .sortedWith { e1: DirectoryEntry, e2: DirectoryEntry -> compareDirectoryEntries(e1, e2) }

    val hash = MessageDigest.getInstance(algName)

    for (entry in entries) {
      if (entry.Name.contentEquals(digitalSignatureEntryName) ||
        (skipMsiDigitalSignatureExEntry && entry.Name.contentEquals(msiDigitalSignatureExEntryName))
      )
        continue

      val data = _cf.GetStreamData(entry)
      hash.update(data)
    }

    val rootClsid = _cf.GetRootDirectoryClsid()

    if (rootClsid != null)
      hash.update(rootClsid)

    return hash.digest()
  }

  private fun compareDirectoryEntries(e1: DirectoryEntry, e2: DirectoryEntry): Int {
    val a = e1.Name
    val b = e2.Name
    val size = Math.min(a.count(), b.count())

    for (i in 0 until size)
      if (a[i] != b[i])
        return (a[i].toInt() and 0xFF) - (b[i].toInt() and 0xFF)

    return a.count() - b.count()
  }
}

File: net/JetBrains.FormatRipper/src/Cf/MsiFile.cs
using System;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using JetBrains.Annotations;

namespace JetBrains.SignatureVerifier.Cf
{
  /// <summary>
  /// MS Windows Installer compound file
  /// </summary>
  public class MsiFile
  {
    private readonly CompoundFile _cf;

    //\u0005DigitalSignature
    private readonly byte[] _digitalSignatureEntryName = { 0x5, 0x0, 0x44, 0x00, 0x69, 0x00, 0x67, 0x00, 0x69, 0x00, 0x74, 0x00, 0x61, 0x00, 0x6C, 0x00, 0x53, 0x00, 0x69, 0x00, 0x67, 0x00, 0x6E, 0x00, 0x61, 0x00, 0x74, 0x00, 0x75, 0x00, 0x72, 0x00, 0x65, 0x00 };

    //\u0005MsiDigitalSignatureEx
    private readonly byte[] _msiDigitalSignatureExEntryName = { 0x5, 0x0, 0x4D, 0x00, 0x73, 0x00, 0x69, 0x00, 0x44, 0x00, 0x69, 0x00, 0x67, 0x00, 0x69, 0x00, 0x74, 0x00, 0x61, 0x00, 0x6C, 0x00, 0x53, 0x00, 0x69, 0x00, 0x67, 0x00, 0x6E, 0x00, 0x61, 0x00, 0x74, 0x00, 0x75, 0x00, 0x72, 0x00, 0x65, 0x00, 0x45, 0x00, 0x78, 0x00 };

    ///  <summary>
    /// Initializes a new instance of the <see cref="T:JetBrains.SignatureVerifier.Cf.MsiFile"></see>
    ///  </summary>
    ///  <param name="stream">An input stream</param>
    ///  <exception cref="PlatformNotSupportedException">Indicates the byte order ("endianness")
    ///  in which data is stored in this computer architecture is not Little Endian.</exception>
    ///  <exception cref="InvalidDataException">If the input stream contains a compound file with wrong structure</exception>
    public MsiFile([NotNull] Stream stream)
    {
      _cf = new CompoundFile(stream);
    }

    /// <summary>
    /// Retrieve the signature data from MSI
    /// </summary>
    public SignatureData GetSignatureData()
    {
      var data = _cf.GetStreamData(_digitalSignatureEntryName);

      if (data == null)
        return SignatureData.Empty;
      return new SignatureData(null, data);
    }

    /// <summary>
    /// Compute hash of MSI structure
    /// </summary>
    /// <param name="algName">Name of the hashing algorithm</param>
    /// <param name="skipMsiDigitalSignatureExEntry">Skip \u0005MsiDigitalSignatureEx entry data when hashing</param>
    /// <exception cref="ArgumentNullException">if algName is null</exception>
    public byte[] ComputeHash([NotNull] string algName, bool skipMsiDigitalSignatureExEntry)
    {
      if (algName == null) throw new ArgumentNullException(nameof(algName));

      var entries = _cf.GetStreamDirectoryEntries();
      entries.Sort(compareDirectoryEntries);

      using var hash = IncrementalHash.CreateHash(new HashAlgorithmName(algName.ToUpper()));

      foreach (var entry in entries)
      {
        if (entry.Name.SequenceEqual(_digitalSignatureEntryName))
          continue;

        if (skipMsiDigitalSignatureExEntry && entry.Name.SequenceEqual(_msiDigitalSignatureExEntryName))
          continue;

        var data = _cf.GetStreamData(entry);
        hash.AppendData(data);
      }

      var rootClsid = _cf.GetRootDirectoryClsid();

      if (rootClsid != null)
        hash.AppendData(rootClsid);

      return hash.GetHashAndReset();
    }

    private int compareDirectoryEntries(DirectoryEntry e1, DirectoryEntry e2)
    {
      var a = e1.Name;
      var b = e2.Name;

      var size = Math.Min(a.Length, b.Length);

      for (var i = 0; i < size; i++)
        if (a[i] != b[i])
          return (a[i] & 0xFF) - (b[i] & 0xFF);

      return a.Length - b.Length;
    }
  }
}